### ISSUE TITLE
Reading request data

### ISSUE THREAD

**Issue by ilyapuchka** [envoy/Ambassador#16]
As readme says to read request data we can use `DataRead.read(input) { data in ... }` in request handler. This does not work for me as reading block is called asynchronously after request handler returns, so I can not use it to build data to send back. If I try to wait for this block without exiting handler connections is getting stuck.

What I want to accomplish is to redirect all requests that are not stubbed to real server (because it's just insane to stub every request in the flow when you need to test one particular case on the screen that is deep inside navigation and app needs to make dozen of requests before reaching it).  Here is my code:

```swift
struct RedirectResponse: WebApp {
    
    // where to redirect
    let host: String
    let scheme: String
    
    func urlRequest(_ environ: [String: Any]) -> URLRequest {
        //reading environment variables and transforming header keys to their initial values....
        return request
    }
    
    public func app(_ environ: [String : Any], startResponse: @escaping ((String, [(String, String)]) -> Swift.Void), sendBody: @escaping ((Data) -> Swift.Void)) {
        var request = urlRequest(environ)
        let input = environ["swsgi.input"] as! SWSGIInput
        DataReader.read(input) { data in
       // this block is called only after method returns, which is too late as request was already fired.
       // Waiting for this block to be called before firing request keeps connection stuck and block is never called
            request.httpBody = data 
        }
        // makes request and wait for response in synchronous manner, using `group.wait()`
        let (data, statusCode, contentType, headers) = URLSession.shared.data(with: request)
        //send data from real server back
        DataResponse(statusCode: statusCode, statusMessage: "", contentType: contentType, headers: headers, handler: { (_, sendData) in
            sendData(data)
        }).app(environ, startResponse: startResponse, sendBody: sendBody)
    }
    
}
```

Then I set this response as `notFoundResponse` for the router.

With that I receive such logs:

```
2017-02-03 13:03:58 +0000 [info] - DefaultHTTPServer: New connection 87A6CC41-5E67-4585-944D-C409FDC9D16A from [::1]:58322
2017-02-03 13:03:58 +0000 [info] - HTTPConnection: [87A6CC41-5E67-4585-944D-C409FDC9D16A] Header parsed, method=Optional("POST"), path="/auth/oauth2/client/access_token", version="HTTP/1.1", headers=[("Host", "localhost:8080"), ("Accept-Encoding", "gzip;q=1.0, compress;q=0.5"), ("mobile-app", "IOS-838"), ("Accept-Language", "en-US;q=1.0, en;q=0.9"), ("Content-Length", "177"), ("Content-Type", "application/json"), ("Accept", "*/*"), ("Connection", "keep-alive"), ("User-Agent", "Alamofire/4.2.0")]
2017-02-03 13:03:58 +0000 [info] - HTTPConnection: [87A6CC41-5E67-4585-944D-C409FDC9D16A] Resume reading
2017-02-03 13:03:58 +0000 [info] - HTTPConnection: [87A6CC41-5E67-4585-944D-C409FDC9D16A] Finish response
2017-02-03 13:03:58 +0000 [info] - HTTPConnection: [87A6CC41-5E67-4585-944D-C409FDC9D16A] Connection closed, reason=byLocal
```

Request fails as it does not contain required body data.

When trying to wait for read data block:

```
2017-02-03 13:01:35 +0000 [info] - DefaultHTTPServer: New connection 6DB9C3DF-02E3-49F5-8801-22EB0CE58AF5 from [::1]:58286
2017-02-03 13:01:35 +0000 [info] - HTTPConnection: [6DB9C3DF-02E3-49F5-8801-22EB0CE58AF5] Header parsed, method=Optional("POST"), path="/auth/oauth2/client/access_token", version="HTTP/1.1", headers=[("Host", "localhost:8080"), ("Accept-Encoding", "gzip;q=1.0, compress;q=0.5"), ("mobile-app", "IOS-838"), ("Accept-Language", "en-US;q=1.0, en;q=0.9"), ("Content-Length", "177"), ("Content-Type", "application/json"), ("Accept", "*/*"), ("Connection", "keep-alive"), ("User-Agent", "Alamofire/4.2.0")]
2017-02-03 13:01:35 +0000 [info] - HTTPConnection: [6DB9C3DF-02E3-49F5-8801-22EB0CE58AF5] Resume reading
```

and then request fails with timeout.

Here is the request that app makes:

```
curl -i \
	-X POST \
	-H "mobile-app: IOS-838" \
	-H "Content-Type: application/json" \
	-H "Accept-Encoding: gzip;q=1.0, compress;q=0.5" \
	-H "Accept-Language: en-US;q=1.0, en;q=0.9" \
	-H "User-Agent: Alamofire/4.2.0" \
	-d "{\"country\":\"us\",\"locale\":\"en-US\",\"grant_type\":\"client_credentials\",\"client_id\":\"...\",\"client_secret\":\"...\",\"scope\":\"public\"}" \
	"http://localhost:8080/auth/oauth2/client/access_token"
```

**Comment by victorlin** [comment]
all operations with Embassy needs to be done in async, you cannot call blocking thing in side the loop, will block it. 

```
struct RedirectResponse: WebApp {

    func urlRequest(_ environ: [String: Any]) -> URLRequest {
        //reading environment variables and transforming header keys to their initial values....
        var request = URLRequest(url: URL(string: "http://httpbin.org/post")!)
        request.httpMethod = "POST"
        return request
    }

    public func app(_ environ: [String : Any], startResponse: @escaping ((String, [(String, String)]) -> Swift.Void), sendBody: @escaping ((Data) -> Swift.Void)) {
        let input = environ["swsgi.input"] as! SWSGIInput
        DataReader.read(input) { incomingData in
            var request = self.urlRequest(environ)
            request.httpBody = incomingData
            let task = URLSession.shared.dataTask(with: request) { (data, resp, error) in
                startResponse("200 OK", [])
                if let data = data {
                    sendBody(data)
                }
                sendBody(Data())
            }
            task.resume()
        }
    }
}
```

Here's simple example I did for trying to get what you want. and the outcome

```
$ curl -i -X POST -H "mobile-app: IOS-838" -H "Content-Type: application/json" -H "Accept-Encoding: gzip;q=1.0, compress;q=0.5" -H "Accept-Language: en-US;q=1.0, en;q=0.9" -H "User-Agent: Alamofire/4.2.0" -d "{\"country\":\"us\",\"locale\":\"en-US\",\"grant_type\":\"client_credentials\",\"client_id\":\"...\",\"client_secret\":\"...\",\"scope\":\"public\"}" "http://localhost:8080/auth/oauth2/client/access_token"
HTTP/1.1 200 OK
Connection: close
Server: Embassy

{
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {
    "{\"country\":\"us\",\"locale\":\"en-US\",\"grant_type\":\"client_credentials\",\"client_id\":\"...\",\"client_secret\":\"...\",\"scope\":\"public\"}": ""
  }, 
  "headers": {
    "Accept": "*/*", 
    "Accept-Encoding": "gzip, deflate", 
    "Accept-Language": "en-us", 
    "Content-Length": "124", 
    "Content-Type": "application/x-www-form-urlencoded", 
    "Host": "httpbin.org", 
    "User-Agent": "xctest (unknown version) CFNetwork/807.2.14 Darwin/16.4.0 (x86_64)"
  }, 
  "json": null, 
  "origin": "xxxxx", 
  "url": "http://httpbin.org/post"
}
```

**Comment by ilyapuchka** [comment]
@victorlin thanks for the answer, that's exactly what I did in the beginning, but requests were never finishing. Now I see that I was missing calling `sendBody` with empty data (I put it in else branch).
But with that I have random crashes on SelectorEventLoop.swift:209 (EXC_BAD_INSTRUCTION) after request returns. Sometimes in happens in other places. Also trying to serialise response I got error "Garbage at end." from time to time.
Here is my code: https://gist.github.com/ilyapuchka/c9c3efb405e30207e6dd3003df3ec4c9
